src/backend/farmer/direct_node_client.rs

```diff
@@ -0,0 +1,606 @@
+                        )));
+
+                    maybe_archived_segment_header
+                };
+
+                Box::pin(async move { maybe_archived_segment_header })
+            });
+
+        let archived_segment_acknowledgement_senders =
+            self.archived_segment_acknowledgement_senders.clone();
+        let fut = async move {
+            let mut archived_segment_acknowledgement_senders =
+                archived_segment_acknowledgement_senders.lock();
+
+            archived_segment_acknowledgement_senders
+                .senders
+                .remove(&subscription_id);
+        };
+
+        self.subscription_executor.spawn(
+            "subspace-archived-segment-header-subscription",
+            Some("rpc"),
+            fut.boxed(),
+        );
+
+        Ok(Box::pin(stream))
+    }
+
+    async fn acknowledge_archived_segment_header(
+        &self,
+        segment_index: SegmentIndex,
+    ) -> Result<(), Error> {
+        let archived_segment_acknowledgement_senders =
+            self.archived_segment_acknowledgement_senders.clone();
+
+        let maybe_sender = {
+            let mut archived_segment_acknowledgement_senders_guard =
+                archived_segment_acknowledgement_senders.lock();
+
+            (archived_segment_acknowledgement_senders_guard.segment_index == segment_index)
+                .then(|| {
+                    let last_key = *archived_segment_acknowledgement_senders_guard
+                        .senders
+                        .keys()
+                        .next()?;
+
+                    archived_segment_acknowledgement_senders_guard
+                        .senders
+                        .remove(&last_key)
+                })
+                .flatten()
+        };
+
+        if let Some(sender) = maybe_sender {
+            if let Err(error) = sender.unbounded_send(()) {
+                if !error.is_closed() {
+                    warn!("Failed to acknowledge archived segment: {error}");
+                }
+            }
+        }
+
+        debug!(%segment_index, "Acknowledged archived segment.");
+
+        Ok(())
+    }
+
+    async fn piece(&self, piece_index: PieceIndex) -> Result<Option<Piece>, Error> {
+        let archived_segment = {
+            let mut cached_archived_segment = self.cached_archived_segment.lock();
+
+            match cached_archived_segment
+                .as_ref()
+                .and_then(CachedArchivedSegment::get)
+            {
+                Some(archived_segment) => archived_segment,
+                None => {
+                    if piece_index > SegmentIndex::ZERO.last_piece_index() {
+                        return Ok(None);
+                    }
+
+                    debug!(%piece_index, "Re-creating genesis segment on demand");
+
+                    // Try to re-create genesis segment on demand
+                    match recreate_genesis_segment(&*self.client, self.kzg.clone()) {
```

Comment:
This is problematic in async context, please wrap it in `spawn_blocking` and await here. In RPC it was a blocking RPC endpoint and didn't need to be wrapped because of it.

Reply:
It is 100% `Send` and `Sync`. Just make sure you move a clone of `Kzg` and not moving `self`